### PR TITLE
[Add] Logits Processor Zoo with HF Transformers

### PR DESCRIPTION
**Pull request by ariG23498**
Adding a blog post on logits processor zoo from NVIDIA and its usage with Hugging Face transformers.

### CODE REVIEW

**Review by Vaibhavs10 — COMMENTED** [review]
Great job on putting this together, it already reads well. I added some suggestion/ nits.

File: logits-processor-zoo.md [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+            print(f"Prompt: {prompt}")
+            print()
+            print(f"LLM response:\n{generated_text}")
+
+runner = LLMRunner()
+```
+
+### 1\. **GenLengthLogitsProcessor**
+
+Control the length of generated sequences by adjusting the likelihood of the end-of-sequence (EOS) token.
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply logits-processor-zoo.md:143]: Very nice suggestion. Will update.

File: logits-processor-zoo.md [hunk 2]
```diff
@@ -0,0 +1,378 @@
+---
+title: "Logit Processing in Language Models: Supercharge Your LLMs with NVIDIA's LogitsProcessorZoo"
```
Comment (Vaibhavs10) [inline logits-processor-zoo.md:2]: I wonder if we can name it a bit more general - `Controlled Generation in LLMs: Supercharge Your LLMs with NVIDIA's LogitsProcessorZoo`

Or something a bit more generic for ex

**Review by Vaibhavs10 — APPROVED** [review]
LGTM! Thanks for incorporating the suggestions:

Two final suggestions:
1. Put a more generic title - if possible.
2. After each generation, add a line that tells people what to look for specifically in that generation and how it relates to the LogitProcessor.

**Comment by ariG23498** [comment]
@Vaibhavs10 thank you for the review! I have updated the blog according to your review.

**Review by Vaibhavs10 — APPROVED** [review]
Thank you! <3

**Comment by ariG23498** [comment]
Waiting for the NVIDIA team to review. After that is done we can merge this.

File: _blog.yml [hunk 3]
```diff
@@ -5169,3 +5169,12 @@
     - community
     - chemistry
     - science
+
+- local: logits-processor-zoo
+  title: "Controlling Language Model Generation with NVIDIA's LogitsProcessorZoo"
+  author: ariG23498
+  thumbnail: /blog/assets/logits-processor-zoo/thumbnail.png
+  date: December 11, 2024
```
Comment (pcuenca) [inline _blog.yml:5177]: Reminder to update

File: logits-processor-zoo.md [hunk 4]
```diff
@@ -0,0 +1,383 @@
+---
+title: "Controlling Language Model Generation with NVIDIA's LogitsProcessorZoo"
+thumbnail: /blog/assets/logits-processor-zoo/thumbnail.png
+authors:
+- user: ariG23498
+- user: aerdem
+  guest: true
+  org: nvidia
```
Comment (pcuenca) [inline logits-processor-zoo.md:8]: This author profile does not seem to belong to the nvidia org.

File: logits-processor-zoo.md [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+---
+
+# Controlling Language Model Generation with NVIDIA's LogitsProcessorZoo
+
+> **Struggling to get language models to follow your instructions?**
+
+NVIDIA's [LogitsProcessorZoo](https://github.com/NVIDIA/logits-processor-zoo/tree/main) provides powerful tools for controling the behavior of language models during text generation. Whether you want to control sequence lengths, enforce specific phrases, or guide multiple-choice answers, this library offers precise control over model outputs. In this post, we'll dive into its features and show how you can use it to refine your AI workflows.
```
Comment (pcuenca) [inline logits-processor-zoo.md:15]:
```suggestion
NVIDIA's [LogitsProcessorZoo](https://github.com/NVIDIA/logits-processor-zoo/tree/main) provides powerful tools for controlling the behavior of language models during text generation. Whether you want to control sequence lengths, enforce specific phrases, or guide multiple-choice answers, this library offers precise control over model outputs. In this post, we'll dive into its features and show how you can use it to refine your AI workflows.
```

File: logits-processor-zoo.md [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+# Decode logits to generate text
+predicted_token_ids = torch.argmax(next_token_probs, dim=-1)
+generated_text = tokenizer.batch_decode(predicted_token_ids, skip_special_tokens=True)
+print("Generated Text:", generated_text[0])
+
+>>> Generated Text: Paris
+```
+
+While this pipeline works for general text generation, **raw logits are not optimized for task-specific constraints or custom behaviors**. This is where logit processing becomes indispensable.
```
Comment (pcuenca) [inline logits-processor-zoo.md:62]: Maybe mention somewhere that these transformations are automatically performed by the transformers pipelines and the `generate()` method.

File: logits-processor-zoo.md [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+- **Overgeneralization:** The model could prioritize generic responses instead of specific, high-quality outputs.
+- **Task misalignment:** Sequences may end too early, be overly verbose, or miss critical details.
+
+Logit processing enables us to tweak the model's behavior by modifying these raw scores before generation.
+
+## NVIDIA's LogitsProcessorZoo
+
+NVIDIA's [LogitsProcessorZoo](https://github.com/NVIDIA/logits-processor-zoo) simplifies post-processing of logits with modular components tailored for specific tasks.
+Let's explore its features and see how to use them. To follow along, head over to
+[the notebook](https://huggingface.co/datasets/ariG23498/quick-notebooks/blob/main/processors.ipynb) and experiment with the logit processors.
```
Comment (pcuenca) [inline logits-processor-zoo.md:78]:
```suggestion
[the notebook](https://huggingface.co/datasets/ariG23498/quick-notebooks/blob/main/nvidia-logits-processor-zoo.ipynb) and experiment with the logits processors.
```

File: logits-processor-zoo.md [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+Install the library using:
+
+```bash
+pip install logits-processor-zoo
+```
+
+To demonstrate the processors, we'll create a simple `LLMRunner` class that initializes a model and tokenizer, exposing a `generate_response` method.
+
+```python
+# Adapted from: https://github.com/ariG23498/logits-processor-zoo/blob/main/example_notebooks/transformers/utils.py
```
Comment (pcuenca) [inline logits-processor-zoo.md:89]: Is there a reason to use your fork instead of nvidia's?

File: logits-processor-zoo.md [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+class LLMRunner:
+    def __init__(self, model_name="meta-llama/Llama-3.2-1B-Instruct"):
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+        self.model = AutoModelForCausalLM.from_pretrained(
+            model_name,
+            torch_dtype=torch.float16,
```
Comment (pcuenca) [inline logits-processor-zoo.md:96]: nit: why not bfloat16, as recommended for this particular model and as used in the first code snippet of the post?

File: logits-processor-zoo.md [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+Let's explore its features and see how to use them. To follow along, head over to
+[the notebook](https://huggingface.co/datasets/ariG23498/quick-notebooks/blob/main/processors.ipynb) and experiment with the logit processors.
+
+Install the library using:
+
+```bash
+pip install logits-processor-zoo
+```
+
+To demonstrate the processors, we'll create a simple `LLMRunner` class that initializes a model and tokenizer, exposing a `generate_response` method.
```
Comment (pcuenca) [inline logits-processor-zoo.md:86]: I'd maybe mention that we need to supply a list of processors to control generation.

File: logits-processor-zoo.md [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+            ].strip()
+
+            print(f"Prompt: {prompt}")
+            print()
+            print(f"LLM response:\n{generated_text}")
+
+runner = LLMRunner()
+```
+
+### 1. **GenLengthLogitsProcessor**
```
Comment (pcuenca) [inline logits-processor-zoo.md:144]: Some breadcrumbs before enumerating the available processors could be helpful, reminding that we can combine them in our calls to the previous class.

File: logits-processor-zoo.md [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+
+The next morning, the old woman helped Max find his way back to the path he had taken. As they said their goodbyes, the old woman handed Max a small gift – a small wooden acorn with a note that read, "Remember, the forest is full of magic, but it's also full of kindness. Always look for it."
+
+Max returned home, forever changed by his experience in the forest. He never forgot the kindness of the old woman and the magic of the forest, and he always carried the lessons he learned with him.
+
+From that day on, Max became known as the forest explorer, and his love for the outdoors was only matched by his love for the people he met along the way. And whenever he looked up at the stars, he remembered the wise words of the old woman: "The forest is full of magic, but it's also full of kindness. Always look for it."
+```
+
+In the examples above, we have used the `GenLengthLogitsProcessor` to both shorten and lengthen the 
+our generated response from the model.
```
Comment (pcuenca) [inline logits-processor-zoo.md:208]:
```suggestion
response generated by the model.
```

File: logits-processor-zoo.md [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+    example_prompts,
+    [GenLengthLogitsProcessor(runner.tokenizer, boost_factor=-10.0, p=0, complete_sentences=False)]
+))
+
+>>> Prompt: Tell me a story about a kid lost in forest.
+
+LLM response:
+Once upon a time, in a dense and vibrant forest, there lived a young boy named Max. Max was an adventurous and curious 8-year-old who loved exploring the outdoors. One sunny afternoon, while wandering through the forest, he stumbled upon a narrow path he had never seen before.
+
+Excited by the discovery, Max decided to follow the path and see where it would lead. The forest was teeming with life, and the sunlight filtering through the trees created a magical atmosphere. Max walked for about 20 minutes, his eyes scanning the surroundings for any signs of civilization.
```
Comment (pcuenca) [inline logits-processor-zoo.md:186]: I'm not sure how the text would render inside a `python` code block. Maybe use quotes instead.

File: logits-processor-zoo.md [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+
+```py
+example_prompts =[
+    "Tell me a story about a kid lost in forest."
+]
+
+# generate short response
+print(runner.generate_response(
+    example_prompts,
+    [GenLengthLogitsProcessor(runner.tokenizer, boost_factor=0.1, p=2, complete_sentences=True)]
```
Comment (pcuenca) [inline logits-processor-zoo.md:161]: A short mention of what `boost_factor` and `p` mean could be helpful.

File: logits-processor-zoo.md (hunk 11 above)
Comment (pcuenca) [inline logits-processor-zoo.md:144]:
```suggestion
### 1. GenLengthLogitsProcessor
```

File: logits-processor-zoo.md [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+
+Max returned home, forever changed by his experience in the forest. He never forgot the kindness of the old woman and the magic of the forest, and he always carried the lessons he learned with him.
+
+From that day on, Max became known as the forest explorer, and his love for the outdoors was only matched by his love for the people he met along the way. And whenever he looked up at the stars, he remembered the wise words of the old woman: "The forest is full of magic, but it's also full of kindness. Always look for it."
+```
+
+In the examples above, we have used the `GenLengthLogitsProcessor` to both shorten and lengthen the 
+our generated response from the model.
+
+### 2. **CiteFromPromptLogitsProcessor**
```
Comment (pcuenca) [inline logits-processor-zoo.md:210]:
```suggestion
### 2. CiteFromPromptLogitsProcessor
```

File: logits-processor-zoo.md [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+    example_prompts,
+    [CiteFromPromptLogitsProcessor(runner.tokenizer, example_prompts, boost_factor=5.0)],
+    max_tokens=50,
+))
+
+>>> Prompt: 
+    A user review: very soft, colorful, expensive but deserves its price, stylish.
+    
+    What is the user's opinion about the product's price?
+
```
Comment (pcuenca) [inline logits-processor-zoo.md:240]: I'd remove the prompt repetition here for brevity.

File: logits-processor-zoo.md [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+    What is the user's opinion about the product's price?
+    
+
+LLM response:
+Based on the user review, the user's opinion about the product's price is: the user is very satisfied, but the price is expensive, but the product is stylish, soft, and colorful, which is the price the user is willing to pay
+```
+
+Notice how the generation cites the input prompt.
+
+### 3. **ForceLastPhraseLogitsProcessor**
```
Comment (pcuenca) [inline logits-processor-zoo.md:248]:
```suggestion
### 3. ForceLastPhraseLogitsProcessor
```

File: logits-processor-zoo.md [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+phrase = "\n\nReferences:"
+batch_size = len(example_prompts)
+
+print(runner.generate_response(
+    example_prompts,
+    [ForceLastPhraseLogitsProcessor(phrase, runner.tokenizer, batch_size)]
+))
+
+
+>>> Prompt:
```
Comment (pcuenca) [inline logits-processor-zoo.md:280]: same as before

File: logits-processor-zoo.md [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+    What is Bulbasaur?
+    
+
+LLM response:
+Bulbasaur is a fictional Pokémon species in the Pokémon franchise. It is a Grass and Poison-type Pokémon, characterized by its distinctive appearance.
+
+Thanks for trying our RAG application! If you have more questions about Bulbasaur, feel free to ask.
+```
+
+With each generation we were able to add the `phrase` string right at the end of the generation.
```
Comment (pcuenca) [inline logits-processor-zoo.md:330]:
```suggestion
With each generation we were able to add the `phrase` string right before the end of the generation.
```

(it's still autocompleted)

File: logits-processor-zoo.md [hunk 20]
```diff
@@ -0,0 +1,383 @@
...
+
+### 4. **MultipleChoiceLogitsProcessor**
```
Comment (pcuenca) [inline logits-processor-zoo.md:332]:
```suggestion
### 4. MultipleChoiceLogitsProcessor
```

File: logits-processor-zoo.md [hunk 21]
```diff
@@ -0,0 +1,383 @@
...
+
+Answer:
+
+
+LLM response:
+1
+```
+
+Here our model does not generate anything other than the choice. This is an immensely helpful attribute while
+working with agnets or using models for multiple choice questions.
```
Comment (pcuenca) [inline logits-processor-zoo.md:375]:
```suggestion
working with agents or using models for multiple choice questions.
```

File: logits-processor-zoo.md [hunk 22]
```diff
@@ -0,0 +1,383 @@
...
+
+## Hugging Face's LogitsProcessor API
+
+While NVIDIA's LogitsProcessorZoo is powerful, it's worth mentioning Hugging Face's [LogitsProcessor API](https://huggingface.co/docs/transformers/en/internal/generation_utils#logitsprocessor). Explore the `transformers` documentation for more details and examples.
```
Comment (pcuenca) [inline logits-processor-zoo.md:379]: This is useful contextualization that, in my opinion, should appear at the beginning of the post to frame the rest of the content. So the idea would be to summarize that selecting the most probable token (aka greedy search) is the obvious approach, but there are many other strategies, [like the ones listed here](https://huggingface.co/docs/transformers/en/generation_strategies) that everybody can use with transformers. Nvidia's toolkit can provide additional options for specific use cases.

File: logits-processor-zoo.md [hunk 23]
```diff
@@ -0,0 +1,383 @@
...
+
+## Wrapping Up
+
+Logit processing is a powerful tool to refine and customize language model outputs. With NVIDIA's **LogitsProcessorZoo** and Hugging Face's `transformers` processors, you can tailor your model's behavior to meet specific requirements. Ready to take your LLMs to the next level? Install these tools and start experimenting today!
```
Comment (pcuenca) [inline logits-processor-zoo.md:383]: Maybe also link to some other resources, like structured generation solutions, for readers that want to dive deeper?

**Review by pcuenca — COMMENTED** [review]
I think some additional contextualization could be helpful, as expressed in a comment.

File: logits-processor-zoo.md (hunk 8 above)
Comment (ariG23498) [reply logits-processor-zoo.md:89]: We are currently working on a [PR](https://github.com/NVIDIA/logits-processor-zoo/pull/2) with which we will be able to use `nvidia`'s repo.

File: logits-processor-zoo.md (hunk 6 above)
Comment (gante) [reply logits-processor-zoo.md:62]: I would also mention that raw logits are also undesirable for common tasks like sampling, linking to our [old generation blog post](https://huggingface.co/blog/how-to-generate)

File: logits-processor-zoo.md [hunk 24]
```diff
@@ -0,0 +1,383 @@
...
+            ]
+            inputs = self.tokenizer.apply_chat_template(
+                conversation,
+                tokenize=True,
+                add_generation_prompt=True,
+                return_tensors="pt",
+                return_dict=True,
+            ).to(self.model.device)
+
+            with torch.inference_mode():
```
Comment (gante) [inline logits-processor-zoo.md:117]:
```suggestion
```
`generate` is decorated with `@torch.no_grad()`, no need for this line :)

**Review by gante — COMMENTED** [review]
+1 to Pedro's [contextualization comment](https://github.com/huggingface/blog/pull/2524/files#r1881989983)

Skimming the top, it gives the impression that `transformers` doesn't include advanced logits processors (which is obviously not true 😛 ). Instead, I would try to frame Nvidia's library as an expansion to the existing processors, and would also highlight that `generate` is compatible with custom logits processors like Nvidia's

**Comment by code30x58** [comment]
https://youtu.be/tKCQcM5vfac?feature=shared

File: logits-processor-zoo.md (hunk 9 above)
Comment (ariG23498) [reply logits-processor-zoo.md:96]: Thanks for the catch. I missed the `b` 😅

File: logits-processor-zoo.md (hunk 24 above)
Comment (ariG23498) [reply logits-processor-zoo.md:117]: Thank you! I have updated in the latest commit.

**Comment by ariG23498** [comment]
@pcuenca @gante thank you for the thorough reviews. I have made the updates, could you all review the changes and let me know if the contextualization is better in this version of the blog post?

File: logits-processor-zoo.md [hunk 25]
```diff
@@ -0,0 +1,343 @@
...
+title: "Controlling Language Model Generation with NVIDIA's LogitsProcessorZoo"
+thumbnail: /blog/assets/logits-processor-zoo/thumbnail.png
+authors:
+- user: ariG23498
+- user: aerdem4
+  guest: true
+  org: nvidia
+---
+
+# Controlling Language Model Generation with NVIDIA's LogitsProcessorZoo
```
Comment (pcuenca) [inline logits-processor-zoo.md:11]: I love how the intro provides great context now, thank you! 🙌

File: logits-processor-zoo.md [hunk 26]
```diff
@@ -0,0 +1,343 @@
...
+
+# Decode logits to generate text
+predicted_token_ids = torch.argmax(next_token_probs, dim=-1)
+generated_text = tokenizer.batch_decode(predicted_token_ids, skip_special_tokens=True)
+print("Generated Text:", generated_text[0])
+
+>>> Generated Text: Paris
+```
+
+While this pipeline demonstrates how raw logits can be transformed into text, it's worth noting that 🤗 Transformers streamline this process.
```
Comment (pcuenca) [inline logits-processor-zoo.md:77]:
```suggestion
While this pipeline demonstrates how raw logits can be transformed into text, it's worth noting that 🤗 Transformers streamlines this process.
```

Transformers (the library) is frequently used in singular as far as I can tell.

**Review by pcuenca — APPROVED** [review]
Checked the intro and skimmed through the rest, given that several of us had already reviewed. The context and framing look great to me now!

**Pull request merged**